### PR TITLE
Improve NonZero on CUDA/ROCM

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: improve NonZero on CUDA/ROCM.

Without the change, ORT is a lot slower than PT. 

image

With the changes, we see 2% ~ 17% latency reduction compared with stock PT on different kind of data.  

image

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Review by weixingzhang — DISMISSED** [review]

File: onnxruntime/core/providers/cuda/tensor/nonzero_impl.cu [hunk 1]
```diff
@@ -82,9 +116,15 @@ cudaError_t NonZeroOutputPositions(
     cudaStream_t stream, const InputT* x, int64_t x_size, int x_rank, const TArray<fast_divmod>& x_strides,
     const int* prefix_counts, int nonzero_elements, int64_t* results) {
   int num_blocks = NonZeroCalcBlockCount(x_size);
+  if (x_rank > MAX_DIMS) {
   NonZeroOutputPositionsKernel<InputT, NONZERO_THREADS_PER_BLOCK><<<num_blocks, NONZERO_THREADS_PER_BLOCK, 0, stream>>>(
```
Comment (mindest) [inline onnxruntime/core/providers/cuda/tensor/nonzero_impl.cu:120]: nit: indentation

**Comment by mindest** [comment]
Are data in the tables from CUDA? Does ROCm have similar improvement?

**Review by mindest — DISMISSED** [review]

**Comment by pengwa** [comment]
> Are data in the tables from CUDA? Does ROCm have similar improvement?

There were some data on ROCM, but I did not include here. On ROCM the initial gap is small between ORT and PyTorch, so the gain is not that big.

**Review by baijumeswani — APPROVED** [review]

**Comment by pengwa** [comment]
Thanks @weixingzhang  @mindest  @baijumeswani for the review!!

**Pull request merged**